keanu-project/src/main/java/io/improbable/keanu/templating/SequenceItem.java

```diff
@@ -54,7 +54,11 @@ public SequenceItem() {
             throw new IllegalArgumentException("Key " + label + " already exists");
         }
         contents.put(label, v);
-        v.setLabel(label);
+        if (v instanceof ProxyVertex) {
+            ((ProxyVertex) v).addScopeToLabel(getUniqueName());
```

Comment:
I don't like the idea of complicating the `Proxy` interface with this additional method. I understand you've put it here because you don't want to allow `setLabel` on a proxy. Instead I think it would be reasonable to replace your logic

        if (this.getLabel() != null) {
            throw new RuntimeException("You should not set the label on a Proxy Vertex");
        }

with

        if (this.getLabel() != null && !this.getLabel().getUnqualifiedName().equals(label.getUnqualifiedName())) {
            throw new RuntimeException("You should not set the label on a Proxy Vertex");
        }

Reply:
Nice. I didn't want to add a method but couldn't think of a better way of checking. This is much better.